locations/models.py

```diff
@@ -1,14 +1,163 @@
 #!/usr/bin/env python
+# -*- coding: utf-8 -*-
 # vim: ai ts=4 sts=4 et sw=4
 import networkx as nx
 from django.conf import settings
 from django.core.cache import cache, caches, InvalidCacheBackendError
-from django.db import models
+from django.db import connection, models
 from mptt.models import MPTTModel, TreeForeignKey
 
 from common.constants import DATA_SOURCES
 
 
+# Sourced from https://github.com/confirm/django-fulltext-search
+class SearchQuerySet(models.query.QuerySet):
+    '''
+    QuerySet which supports MySQL and MariaDB full-text search.
+    '''
+
+    def __init__(self, fields=None, **kwargs):
+        super(SearchQuerySet, self).__init__(**kwargs)
+        self.search_fields = fields
+
+    def get_query_set(self, query, columns, mode):
+        '''
+        Returns the query set for the columns and search mode.
+        '''
+        # Create the WHERE MATCH() ... AGAINST() expression.
+        fulltext_columns = ', '.join(columns)
+        where_expression = ('MATCH({}) AGAINST("%s" {})'.format(fulltext_columns, mode))
+
+        # Get query set via extra() method.
+        return self.extra(where=[where_expression], params=[query])
```

Comment:
This looks unsafe even though it is expected that portions of it are defined in code. Please see: https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html